compare_platforms/compare_platforms_read_only.c

```diff
@@ -0,0 +1,209 @@
+        assert((min(realname, permissions) + smsz) <= (max(realname, permissions)));
+        return 0;
+    } 
+    else if(max(max(realname, publicreview), permissions) == publicreview)
+    {
+        assert((max(realname, permissions) + smsz) <= publicreview);
+        assert((min(realname, permissions) + smsz) <= (max(realname, permissions)));
+        return 2;
+    }
+    else 
+    {
+        assert((min(realname, permissions) + smsz) <= publicreview);
+        assert((publicreview + bigsz) <=  (max(realname, permissions)));
+        return 1; 
+    }
+}
+
+void free_pointers(struct review *rv)
+{
+    free(rv->realname);
+    free(rv->publicreview);
+    free(rv->permissions);
+}
+
+void rewrite_permissions_by_overflow(struct review *rv, size_t smsz, size_t bigsz, bool bWeak)
+{
+    int revpos = assert_no_overlap(rv->realname, rv->publicreview, rv->permissions, smsz, bigsz);
+    // realname has to come before permissions for the following to make sense
+    assert((rv->realname + smsz) <= rv->permissions);
+        
+    if((revpos == 1) && (rv->publicreview+bigsz <= rv->permissions)) 
+    {
+        // if publicreview is in the middle and permissions is on the right of it
+        assert((rv->publicreview + bigsz + bigsz/2) > rv->permissions);
+        printf("\nOverflowing publicreview into permissions by 1:\n");
+        overflow_into_permissions(bigsz+1, 1, rv->publicreview, rv);
+    }
+    else 
+    {
+        assert((rv->realname + smsz + smsz/2) > rv->permissions);
+        printf("\nOverflowing realname into permissions by 1:\n");
+        overflow_into_permissions(smsz+1, 1, rv->realname, rv);
+    }
+    
+    const size_t oversz = rv->permissions - rv->realname + 2 - smsz;
+    printf("\nOverflowing realname into permissions by %zu:\n", oversz);
+    overflow_into_permissions(smsz+2, oversz, rv->realname, rv);
+}
+
+int main(int argc, char* argv[])
+{
+    struct review review;
+    // In a non-CHERI environment, write protection is weakly enforced
+    bool bWeak = true;
+    
+    // initialization values for struct member sizes
+    const size_t smallsz = 0x20;
+    const size_t biggersz = 0x1001;
+       
+    char *realname = malloc(smallsz);
+    strcpy(realname, "Baba Yaga");
+    review.realname = realname;
+    
+    char *publicreview = malloc(biggersz);
+    strcpy(publicreview, "I am a little unclear as to the contribution. I think the authors could strengthen their case considerably if they conducted an RCT. Weak reject.");
+    review.publicreview = publicreview;
+    
+    char *permissions = malloc(smallsz);
+    strcpy(permissions, "r");
+    review.permissions = permissions;
+    
+    print_details(&review);
+    
+    bool b_improved = false;
+    
+    char *newpublicreview = malloc(biggersz);
+    strcpy(newpublicreview, "5 ***** strong accept. This author deserves two Nobels and an ice cream.\n");
+
+#ifdef __CHERI_PURE_CAPABILITY__    
+    if(can_write(&review) == false) 
```

Comment:
What does this guard check against? Or more specifically, does the value of the boolean ever change? If so, how does that relate to the steps in the preamble?

Reply:
can_write() checks if the review's permissions have a 'w' in them. 
The value of the boolean changes if rewrite_permissions_by_overflow() succeeds in adding one or more 'w's to the permissions.
Then the text of the review can be changed.

Reply:
Right, but I see `rewrite_permissions_by_overflow` is not being called until this point.

Reply:
Sorry, I didn't properly answer the question about how it relates to the steps in the preamble. On line 164, permissions are set to "r". Then we reach this CHERI block where on line 180, it checks the review struct's permissions before setting them to read-only.

With this program flow, the permissions are only ever going to be "r", so arguably this check is not necessary for the purpose of an example. Shall I remove it?

Reply:
Yes, if the check doesn't do anything, best to remove it.

Reply:
Fixed in e55ffa2458505e0a1149a60fe0a421e4da07bca7